wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/devfile/DevfileManager.java

```diff
@@ -121,23 +187,85 @@ public void resolveReference(DevfileImpl devfile, FileContentProvider fileConten
     }
   }
 
-  private DevfileImpl parse(String content, ValidationFunction validationFunction)
+  private DevfileImpl parse(
+      String content, ObjectMapper mapper, Map<String, String> overrideProperties)
       throws DevfileFormatException {
-    JsonNode parsed = validationFunction.validate(content);
-
     DevfileImpl devfile;
     try {
-      devfile = objectMapper.treeToValue(parsed, DevfileImpl.class);
+      JsonNode parsed = mapper.readTree(content);
+      parsed = applyPropertiesOverride(parsed, overrideProperties);
+      schemaValidator.validate(parsed);
+      devfile = mapper.treeToValue(parsed, DevfileImpl.class);
     } catch (JsonProcessingException e) {
       throw new DevfileFormatException(e.getMessage());
+    } catch (IOException e) {
+      throw new DevfileFormatException("Unable to parse Devfile. Error: " + e.getMessage());
     }
-
     integrityValidator.validateDevfile(devfile);
     return devfile;
   }
 
-  @FunctionalInterface
-  private interface ValidationFunction {
-    JsonNode validate(String content) throws DevfileFormatException;
+  private JsonNode applyPropertiesOverride(
```

Comment:
I believe it may be designed in a quite general style, so it could leave in its own Java class. Then DevfileManager and its test class would be simpler

Reply:
Splitted